simpl/middleware/errors.py

```diff
@@ -0,0 +1,192 @@
+
+    """Format outgoing exceptions.
+
+    Uses and is compatible-with bottle exception formatting.
+
+    - Handle Bottle Exceptions (even when catchall=False).
+    - Handle SimplHTTPError
+    - Fail-safe to a generic error (unexpected_error)
+
+    The code in this middleware is meant to be generic.
+    Don't catch and translate fancy exceptions here, do
+    it in the application logic or a separate middleware and
+    raise a bottle.HTTPError or a SimplHTTPError.
+    """
+
+    def __init__(self, app, conf=None):
+        """Initialize class."""
+        if _catchall_enabled(app) is True:
+            LOG.warning("Bottle's catchall flag is enabled for your app. "
+                        "The suggested use for this middleware is with "
+                        "catchall disabled.")
+        self.app = app
+        self.config = conf
+        LOG.debug("Added middleware: %s", type(self).__name__)
+
+    def __call__(self, environ, start_response):
+        """Catch exceptions and format them based on config."""
+        try:
+            return self.app(environ, start_response)
+        except bottle.HTTPError as error:
+            LOG.error("Formatting a bottle exception.",
+                      exc_info=error)
+
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+        except simpl_exc.SimplHTTPError as error:
+            LOG.error("Formatting a SimplHTTPError exception.",
+                      exc_info=error)
+            error = bottle.HTTPError(
+                status=error.status_code, body=error.body,
+                exception=error.exception or error,
+                traceback=error.traceback or tb_mod.format_exc())
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+        except Exception as error:  # pylint: disable=W0703
+            LOG.error("Formatting an unexpected exception.",
+                      exc_info=error)
+            error = bottle.HTTPError(
+                status=500, body=rest.UNEXPECTED_ERROR,
+                exception=error,
+                traceback=tb_mod.format_exc())
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+        except:  # noqa
+            LOG.error("Formatting an unknown exception.",
+                      exc_info=True)
+            error = bottle.HTTPError(
+                status=500, body=rest.UNEXPECTED_ERROR,
+                exception=sys.exc_info()[1],
+                traceback=tb_mod.format_exc())
+            rest.format_error_response(error)
+            start_response(error.status_line, error.headerlist)
+            return error
+
+
+if __name__ == '__main__':
+    # pylint: disable=missing-docstring
+    logging.basicConfig()
+    APP = bottle.Bottle(catchall=False)
```

Comment:
I think it's fine to put a test or example usage here, but this code seems specific to the discussion about SystemExit and KeyboardInterrupt. I'm not sure they remain relevant to stay in the code base.